CASMINST-6596 Adhoc NCN configuration for 1.4.2 release (including sysctl values) (#4122)
Cray-HPE/docs-csm: upgrade/1.4.2/README.md

1. [Upload NCN images](#upload-ncn-images)
1. [Upgrade Ceph and stop local Docker registries](#upgrade-ceph-and-stop-local-docker-registries)
1. [Enable `Smartmon` Metrics on Storage NCNs](#enable-smartmon-metrics-on-storage-ncns)

> 1. [Configure NCN nodes without restart](#configure-ncn-nodes-without-restart)

Reviewer: ```suggestion
1. [Configure NCNs without restart](#configure-ncns-without-restart)
```

   /usr/share/doc/csm/scripts/operations/ceph/enable-smart-mon-storage-nodes.sh
   ```

## Configure NCN nodes without restart

This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated sysctl values.

> 1. (`ncn-m001#`) Create a new CFS Configuration Entry for the release

Reviewer 2: This role exists in `ncn-{master,worker,storage}.yml` and `ncn_nodes.yml`, when are those playbooks ran (do we need this step at all)?

Author: We need a named configuration within CFS that references this top level playbook. The most recent push calls this configuration 'ncn_nodes'. It can be anything, and its not particularly meaningful or relevant given this is run once (imperatively) and doesn't need to be referenced thereafter.

```bash
COMMIT=`kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' 2>/dev/null | yq r -j - 2>/dev/null | jq --arg version "$CSM_RELEASE_VERSION" '. [$version].configuration.commit' | tr -d '"'`
echo '{
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "COMMIT",
      "name": "ncn_sysctl",
      "playbook": "ncn_sysctl.yml"
    }
  ]
}' | sed -e "s/COMMIT/$COMMIT/g" > /tmp/ncn_sysctl.yml.json
cray cfs configurations update ncn_sysctl --file /tmp/ncn_sysctl.yml.json
# Cleanup temporary file
rm /tmp/ncn_sysctl.yml.json
```   

> 2. (`ncn-m001#`) Launch CFS to configure NCN sysctl values

Reviewer 2: The spacing is wrong, and the list is actually broken however it is hard to tell since each item is enumerated. Each code block needs to be indented, as you can see in the screenshot that none of the code blocks show as a nested item to the list item.
![image](https://github.com/Cray-HPE/docs-csm/assets/7772179/f6b2ca4b-f9e0-4a0f-9076-9e5493e9bad6)

Author: Thanks; resolving now.

   /usr/share/doc/csm/scripts/operations/ceph/enable-smart-mon-storage-nodes.sh
   ```

> ## Configure NCN nodes without restart

Reviewer: This should be added to the table of contents at the top of the page.

Author: Done.

> This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated sysctl values.

Reviewer: ```suggestion
This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated `sysctl` values.
```

> This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated sysctl values.

Reviewer 2: ```suggestion
This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated `sysctl` values.
```

Reviewer 2: Jinx, you owe me a coke @mharding-hpe .

> 1. (`ncn-m001#`) Create a new CFS Configuration Entry for the release

Reviewer: ```suggestion
1. (`ncn-m001#`) Create a new CFS configuration entry for the release.
```

   ```bash

>    COMMIT=`kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' 2>/dev/null | yq r -j - 2>/dev/null | jq --arg version "$CSM_RELEASE_VERSION" '. [$version].configuration.commit' | tr -d '"'`

Reviewer 3: nit: I think `jq -r` might do what you're accomplishing with that final `tr -d`

Reviewer 2: Jinx, you also owe me a coke @haasken-hpe .

>    COMMIT=`kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' 2>/dev/null | yq r -j - 2>/dev/null | jq --arg version "$CSM_RELEASE_VERSION" '. [$version].configuration.commit' | tr -d '"'`

Reviewer: ```suggestion
   COMMIT=`kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' 2>/dev/null \
       | yq r -j - 2>/dev/null | jq -r --arg version "$CSM_RELEASE_VERSION" '. [$version].configuration.commit'`
```

   echo '{
     "layers": [
       {
         "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
         "commit": "COMMIT",
         "name": "ncn_nodes",
         "playbook": "ncn_nodes.yml"
       }
     ]
   }' | sed -e "s/COMMIT/$COMMIT/g" > /tmp/ncn_nodes.yml.json
   cray cfs configurations update ncn_nodes --file /tmp/ncn_nodes.yml.json
   # Cleanup temporary file
   rm /tmp/ncn_nodes.yml.json
   ```

> 2. (`ncn-m001#`) Imperatively launch CFS against NCN nodes

Reviewer: ```suggestion
1. (`ncn-m001#`) Imperatively launch CFS against NCNs.
```

   ```bash
   cray cfs sessions create --name ncnnodes --configuration-name ncn_nodes

>    kubectl logs -f -n services jobs/`cray cfs sessions describe ncnnodes --format json | jq " .status.session.job" | tr -d '"'` -c ansible

Reviewer 2: If you use `jq -r` you can remove the `| tr -d '"'`.

Author: "Should have been default behavior for JQ, fight me!" ... updating commands.

>    kubectl logs -f -n services jobs/`cray cfs sessions describe ncnnodes --format json | jq " .status.session.job" | tr -d '"'` -c ansible

Reviewer 3: nit: could also probably use `jq -r` instead of `tr -d` here

   ```

> 3. (`ncn-m001#`) Wait for CFS to complete configuration

Reviewer 3: Might be helpful to say what they're looking for here to indicate success.

> 3. (`ncn-m001#`) Wait for CFS to complete configuration

Reviewer: ```suggestion
1. (`ncn-m001#`) Wait for CFS to complete configuration.
```

   /usr/share/doc/csm/scripts/operations/ceph/enable-smart-mon-storage-nodes.sh
   ```

## Configure NCN nodes without restart

This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated sysctl values.

1. (`ncn-m001#`) Create a new CFS Configuration Entry for the 1.4.2 release

```bash

> VCS_USER=$(kubectl get secret -n services vcs-user-credentials --template={{.data.vcs_username}} | base64 --decode)

Reviewer: Ideally we should get the commit ID from the Cray Product Catalog, IMO

Author: Agreed and done.

VCS_PASSWORD=$(kubectl get secret -n services vcs-user-credentials --template={{.data.vcs_password}} | base64 --decode)
git config --global credential.helper store
echo "https://${VCS_USER}:${VCS_PASSWORD}@api-gw-service-nmn.local" > ~/.git-credentials
git config --global user.name “Root User”
git config --global user.email root@m001localhost
cd /tmp
git clone https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git
cd /tmp/csm-config-management
git checkout cray/csm/1.15.13
COMMIT=`git rev-parse HEAD`
echo '{
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "COMMIT",
      "name": "ncn_sysctl",
      "playbook": "ncn_sysctl.yml"
    }
  ]
}' > /tmp/ncn_sysctl.yml.json
sed -i "s/COMMIT/$COMMIT/g" /tmp/ncn_sysctl.yml.json
cray cfs configurations update ncn_sysctl --file /tmp/ncn_sysctl.yml.json
# Cleanup temporary files
rm -r /tmp/csm-config-management /tmp/ncn_sysctl.yml.json
```   

2. (`ncn-m001#`) Launch CFS to configure NCN sysctl values

```bash

> cray cfs sessions create --name ncn_sysctl --configuration-name ncnsysctl

Reviewer: If the NCNs have a desired configuration set already, won't batcher just run that when this finishes and it sees that the node is out of sync? Or is CFS not smart enough to notice that this session changed the state of a component?

Author: Adhoc sessions may update the overall state of a component; if the registered state is different, reconfiguration will occur. This would be fine as long as the configuration registered to the component is idempotent.

Author: Per phonecall discussion, we are going to do this imperatively in 1.4.2 and allow for a full fledged solution in 1.4.3 or later.

   /usr/share/doc/csm/scripts/operations/ceph/enable-smart-mon-storage-nodes.sh
   ```

> ## Configure NCN nodes without restart

Reviewer: ```suggestion
## Configure NCNs without restart
```

> This step will create an imperative CFS session that can be used to configure booted NCN nodes with updated `sysctl` values.

Reviewer: ```suggestion
This step will create an imperative CFS session that can be used to configure booted management NCNs with updated `sysctl` values.
```

1. (`ncn-m001#`) Create a new CFS configuration entry for the release.

   ```bash
   COMMIT=`kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' 2>/dev/null | yq r -j - 2>/dev/null | jq --arg version "$CSM_RELEASE_VERSION" '. [$version].configuration.commit' | tr -d '"'`
   echo '{
     "layers": [
       {
         "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
         "commit": "COMMIT",
         "name": "ncn_nodes",
         "playbook": "ncn_nodes.yml"
       }
     ]
   }' | sed -e "s/COMMIT/$COMMIT/g" > /tmp/ncn_nodes.yml.json
   cray cfs configurations update ncn_nodes --file /tmp/ncn_nodes.yml.json
   # Cleanup temporary file
   rm /tmp/ncn_nodes.yml.json
   ```

> 1. (`ncn-m001#`) Imperatively launch CFS against NCNs.

Reviewer: ```suggestion
1. (`ncn-m001#`) Imperatively launch CFS against management NCNs.
```

   ```bash
   cray cfs sessions create --name ncnnodes --configuration-name ncn_nodes
   kubectl logs -f -n services jobs/`cray cfs sessions describe ncnnodes --format json | jq -r " .status.session.job"` -c ansible
   ```

1. (`ncn-m001#`) Wait for CFS to complete configuration.

   ```bash
   cray cfs sessions describe ncnnodes
   kubectl logs -f -n services jobs/`cray cfs sessions describe ncnnodes --format json | jq -r " .status.session.job"` -c ansible
   ```

> The playbook in question will run to completion with a message indicating success.

Reviewer: ```suggestion
   The playbook in question will run to completion with a message indicating success.
```
